vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -0,0 +1,132 @@
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+
+
+class FP8Config(QuantizationConfig):
+    """Config class for FP8."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "fp8"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [
+            torch.bfloat16, torch.half, torch.float8_e4m3fn, torch.float8_e5m2
+        ]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 90
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return []
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "FP8Config":
+        return cls()
+
+    def get_linear_method(self) -> "Fp8LinearMethod":
+        return Fp8LinearMethod(self)
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class Fp8LinearState(Enum):
+
+    UNINITIALIZED = enum.auto()
+    READY = enum.auto()
+
+
+class Fp8LinearMethod(LinearMethodBase):
+    """Linear method for FP8.
+
+    Args:
+        quant_config: The quantization config.
+    """
+
+    def __init__(self, quant_config: FP8Config):
+        self.quant_config = quant_config
+
+    def create_weights(
+        self,
+        layer: torch.nn.Module,
+        input_size_per_partition: int,
+        output_size_per_partition: int,
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+        **extra_weight_attrs,
+    ):
+        weight = Parameter(torch.empty(output_size_per_partition,
+                                       input_size_per_partition,
+                                       dtype=params_dtype),
+                           requires_grad=False)
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, extra_weight_attrs)
+
+        scale = Parameter(
+            torch.empty(1, dtype=torch.float32),
+            requires_grad=False,
+        )
+        layer.register_parameter("scale", scale)
+        set_weight_attrs(scale, extra_weight_attrs)
+        layer.fp8_linear_state = Fp8LinearState.UNINITIALIZED
+
+    def apply_weights(self,
+                      layer: torch.nn.Module,
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        qinput, scale = per_tensor_quantize(x)
+
+        if layer.fp8_linear_state == Fp8LinearState.UNINITIALIZED:
+            qweight, weight_scale = per_tensor_quantize(layer.weight)
+            layer.weight.data = qweight.t()
+            layer.scale.data = weight_scale
+            layer.fp8_linear_state = Fp8LinearState.READY
+
+        output, _ = torch._scaled_mm(
+            qinput,
+            layer.weight,
+            out_dtype=x.dtype,
+            scale_a=scale,
+            scale_b=layer.scale,
+            bias=bias,
+        )
```

Comment:
Could we add emulated support for devices without FP8 hardware? Something like this where it simply dequantizes:
```python
        cuda_compute_capability = torch.cuda.get_device_capability()
        if cuda_compute_capability >= (9, 0):
            output, _ = torch._scaled_mm(
                qinput,
                layer.weight,
                out_dtype=x.dtype,
                scale_a=scale,
                scale_b=layer.scale,
                bias=bias,
            )
        else:
            # For NVIDIA SM < 9.0
            logger.warn("FP8 hardware support doesn't exist for "
                      "NVIDIA SM < 9.0. Up-conversion to "
                      "original dtype will be used.")

            output = F.linear(
                qinput.to(x.dtype) * scale, 
                layer.weight.to(x.dtype) * layer.scale.to(x.dtype),
                bias)
```

Reply:
Of course this can be done to benefit other GPUs' memory footprint. Meanwhile, the current quantization config already checks the compatibility and throws an error when running on SM <9.0 with `-q fp8`. We could think about a proper interface.